upd: Data Analysis (#5514)
JetBrains/kotlin-web-site: docs/topics/data-analysis/data-analysis-connect-to-db.md

[//]: # (title: Connect and retrieve data from databases)

> [//]: # (description: Learn how to connect to SQL databases from Kotlin Notebook, inspect table schemas, and retrieve data with Kotlin DataFrame.)

Reviewer: Could we please remove "Kotlin Notebook" in naming here

Author: Removed all mentions from the descriptions, except for the dedicated Notebooks page

[Kotlin Notebook](kotlin-notebook-overview.md) provides support for the most common SQL databases:

> * [PostgreSQL](https://kotlin.github.io/dataframe/postgresql.html)

Reviewer 2: I see you've followed the order based on the DataFrame docs. What do you think about alphabetizing this list?

Author: Yes, good point! Updated.

* [MySQL](https://kotlin.github.io/dataframe/mysql.html)
* [Microsoft SQL Server](https://kotlin.github.io/dataframe/microsoft-sql-server.html)
* [SQLite](https://kotlin.github.io/dataframe/sqlite.html)
* [H2](https://kotlin.github.io/dataframe/h2.html)
* [MariaDB](https://kotlin.github.io/dataframe/mariadb.html)
* [DuckDB](https://kotlin.github.io/dataframe/duckdb.html)

With [Kotlin DataFrame library](https://kotlin.github.io/dataframe/home.html), Kotlin Notebook can establish connections to databases, 
execute SQL queries, and import the results for further operations.

> > For a detailed example, see the [Notebook in the KotlinDataFrame SQL Examples GitHub repository](https://github.com/zaleslaw/KotlinDataFrame-SQL-Examples/blob/master/notebooks/imdb.ipynb).

Reviewer 2: ```suggestion
> For a detailed example, explore the notebook in the [KotlinDataFrame SQL Examples GitHub repository](https://github.com/zaleslaw/KotlinDataFrame-SQL-Examples/blob/master/notebooks/imdb.ipynb).
```
We either refer to it as lowercase "notebook" or capitalize the full product name: Kotlin Notebook :)

If the Kotlin Notebook features are not available, ensure the plugin is enabled. For more information,
see [Set up an environment](kotlin-notebook-set-up-env.md).

To follow this tutorial:
1. Create a [new Kotlin Notebook](kotlin-notebook-create.md).

> 2. Add the Java Database Connectivity (JDBS) driver dependency for your database. 

Reviewer 2: Where should this be added in the notebook?

   For example, to connect to a MariaDB database, add:

   ```kotlin 
   USE {
      dependencies("org.mariadb.jdbc:mariadb-java-client:$version")
   }
   ```
3. Import Kotlin DataFrame:
```kotlin
%use dataframe
```

> > Make sure to run the code cell with the `%use dataframe` line before you run any other code cells

Reviewer 2: I know this was here before your work, but can you explain why this is necessary? 🤔

Author: Done

> that rely on the Kotlin DataFrame library.
>
{style="note"}

## Connect to a database

> To connect to a database, create a connection configuration using `DbConnectionConfig`:

Reviewer 2: ```suggestion
To connect to a database, create a connection configuration using the `DbConnectionConfig()` function:
```

> 1. Import required functionality:

Reviewer 2: ```suggestion
1. Add the following imports:
```
It sounds a bit formal. How does this sound?

Author: Combined our versions :)

   
   ```kotlin
   import org.jetbrains.kotlinx.dataframe.io.DbConnectionConfig
   import org.jetbrains.kotlinx.dataframe.schema.DataFrameSchema
   ```

> 2. Define connection parameters (URL, username, password) using `DbConnectionConfig`:

Reviewer 2: ```suggestion
2. Define connection parameters (URL, username, password) using the `DbConnectionConfig()` function:
```

   ```kotlin
   val URL = "YOUR_URL"
   val USER_NAME = "YOUR_USERNAME"
   val PASSWORD = "YOUR_PASSWORD"
   
   val dbConfig = DbConnectionConfig(URL, USER_NAME, PASSWORD)
   ```

> For more information on connecting to SQL databases, see [Read from SQL databases in the Kotlin DataFrame documentation](https://kotlin.github.io/dataframe/readsqldatabases.html).
>
{style="tip"}

## Inspect database schema

> Before [loading the data](#load-data), you can inspect the database schemas. 

Reviewer 2: We don't usually add links to following sections because the reader will naturally progress there. Can you remove them?

> This helps you understand what tables you have and what columns they contain. 

Reviewer 2: ```suggestion
Before [loading the data](#load-data), inspect the database schemas to understand what tables there are and what columns they contain. 
```
What do you think about making this more direct?

> You can also use the schemas to decide which table to load into a `DataFrame`.

Reviewer 2: ```suggestion
You can use the schemas to decide which table to load into a `DataFrame`.
```
If you accept the more direct suggestion above.

> To retrieve schemas for all non-system tables in your database, use `DataFrameSchema.readAllSqlTables()`:

Reviewer 2: What does non-system mean here?

Author: Changed to user-created.

```kotlin
val dataSchemas = DataFrameSchema.readAllSqlTables(dbConfig)

dataSchemas.forEach { (tableName, schema) ->
    println("---Schema for table: $tableName---")
    println(schema)
    println()
}
```

## Load data

After you [inspect the database schemas](#inspect-database-schema) and select the data, load the

> data into a `DataFrame`.

Reviewer 2: ```suggestion
data into a DataFrame.
```

> Kotlin DataFrame provides two main ways to load data from a database:

Reviewer 2: ```suggestion
Kotlin DataFrame provides two ways to load data from a database:
```
If we don't mention alternatives later then I think we should remove "main" here :)

> * Use [`DataFrame.readSqlTable`](https://kotlin.github.io/dataframe/readsqldatabases.html#reading-specific-tables) to load data directly from a table.

Reviewer 2: ```suggestion
* Use the [`DataFrame.readSqlTable()`](https://kotlin.github.io/dataframe/readsqldatabases.html#reading-specific-tables) function to load data directly from a table.
```
We try to describe what the element is so that readers can understand without having to dig into the source code.

> * Use [`DataFrame.readSqlQuery`](https://kotlin.github.io/dataframe/readsqldatabases.html#executing-sql-queries) to load the result of a custom SQL query.

Reviewer 2: ```suggestion
* Use the [`DataFrame.readSqlQuery()`](https://kotlin.github.io/dataframe/readsqldatabases.html#executing-sql-queries) function to load the result of a custom SQL query.
```

> * Use [`DataFrame.readSqlQuery`](https://kotlin.github.io/dataframe/readsqldatabases.html#executing-sql-queries) to load the result of a custom SQL query.

Reviewer 2: To reduce duplication, how about we make this more concise? Then you can add the specific functions in the dedicated sections.

```suggestion
* Load data directly from a table.
* Load the result of a custom SQL query.
```

Author: Yes, it's much better, thanks!

> Both functions return a `DataFrame` that you can inspect, transform, and analyze in Kotlin Notebook.

Reviewer 2: ```suggestion
Both approaches return a DataFrame that you can inspect, transform, and analyze in Kotlin Notebook.
```

### Load data from a table

> To load data from a table, use [`DataFrame.readSqlTable`](https://kotlin.github.io/dataframe/readsqldatabases.html#reading-specific-tables).

Reviewer 2: ```suggestion
To load data from a table, use the [`DataFrame.readSqlTable()`](https://kotlin.github.io/dataframe/readsqldatabases.html#reading-specific-tables) function.
```

> The following example loads the first 100 rows from the `movies`:

Reviewer 2: It's not clear here what 'movies' is referring to? Is it a table? I don't see it in the code sample 🤔

Author: Yes, see line 9 https://github.com/zaleslaw/KotlinDataFrame-SQL-Examples/blob/master/notebooks/imdb.ipynb

```kotlin
val moviesDf = DataFrame.readSqlTable(
    dbConfig = dbConfig,
    tableName = "movies",
    limit = 100
)

moviesDf
```

### Load data with an SQL query

> To execute a specific SQL query on your database, use [`DataFrame.readSqlQuery`](https://kotlin.github.io/dataframe/readsqldatabases.html#executing-sql-queries). 

Reviewer 2: ```suggestion
To execute a specific SQL query on your database, use the [`DataFrame.readSqlQuery()`](https://kotlin.github.io/dataframe/readsqldatabases.html#executing-sql-queries) function. 
```

> This helps when you need to load specific columns, join tables, filter rows, or aggregate data in a database.

Reviewer 2: ```suggestion
This approach is useful when you need to load specific columns, join tables, filter rows, or aggregate data in a database.
```

> Let’s retrieve a specific dataset related to movies directed by Quentin Tarantino. 

Reviewer 2: ```suggestion
Let's retrieve a specific dataset related to movies directed by Quentin Tarantino. 
```

Author: Oh, that's a really nice catch! Thanks

This query selects movie details and combines genres for each movie:

```kotlin
val TARANTINO_FILMS_SQL_QUERY = """
    SELECT name, year, rank, GROUP_CONCAT(genre) as "genres"
    FROM movies JOIN movies_directors ON movie_id = movies.id
    JOIN directors ON directors.id=director_id LEFT JOIN movies_genres ON movies.id = movies_genres.movie_id
    WHERE directors.first_name = "Quentin" AND directors.last_name = "Tarantino"
    GROUP BY name, year, rank
    ORDER BY year
    """

val tarantinoMoviesDf = DataFrame.readSqlQuery(dbConfig, TARANTINO_FILMS_SQL_QUERY)

tarantinoMoviesDf
```

## Process data

After [loading your database](#load-data) into a `DataFrame`, you can use DataFrame
operations to process retrieved data.

> For example, let’s manipulate data from the previous section. The following code:

Reviewer 2: ```suggestion
For example, let's manipulate data from the previous section. The following code:
```

1. Replaces missing values in the `year` column using [`fillNA`](https://kotlin.github.io/dataframe/fill.html#fillna).
2. Converts the column to `Int` using [`convert`](https://kotlin.github.io/dataframe/convert.html).

> 3. Keeps only films released after 2000 using [`filter`](https://kotlin.github.io/dataframe/filter.html).

Reviewer 2: ```suggestion
1. Replaces missing values in the `year` column using the [`fillNA()`](https://kotlin.github.io/dataframe/fill.html#fillna) function.
2. Converts the column to `Int` type using the [`convert()`](https://kotlin.github.io/dataframe/convert.html) function.
3. Keeps only films released after 2000 using the [`filter()`](https://kotlin.github.io/dataframe/filter.html) function.
```

```kotlin
val filteredTarantinoMovies = tarantinoMoviesDf
    .fillNA { year }.with { 0 }
    .convert { year }.toInt()
    .filter { year > 2000 }

filteredTarantinoMovies
```

## Analyze data

Use [Kotlin Notebook](kotlin-notebook-overview.md) and [DataFrame library](https://kotlin.github.io/dataframe/home.html)
to group, sort, and aggregate data. 

> This helps you uncover and understand patterns of your data.

Reviewer 2: ```suggestion
to group, sort, and aggregate data so you can uncover and understand patterns in your data.
```
"In" sounds a bit more natural. I also suggest to connect the two sentences as "this helps" reads a little vague to me what "this" means.

> For example, let’s read actor data from the `actors` table and find the 20 most common first names:

Reviewer 2: ```suggestion
For example, let's read actor data from the `actors` table and find the 20 most common first names:
```

If the Kotlin Notebook features are not available, ensure the plugin is enabled. For more information,
see [Set up an environment](kotlin-notebook-set-up-env.md).

To follow this tutorial:
1. Create a [new Kotlin Notebook](kotlin-notebook-create.md).
2. Add the Java Database Connectivity (JDBC) driver dependency for your database in the first cell of your notebook.
   
   For example, to connect to a MariaDB database, add:

   ```kotlin 
   USE {
      dependencies("org.mariadb.jdbc:mariadb-java-client:$version")
   }
   ```
3. Import Kotlin DataFrame:
```kotlin
%use dataframe

> ```

Reviewer 2: `````suggestion
   ```kotlin
   %use dataframe
   ```
`````
This code block needs to be indented to be in line with the step.

> Make sure to run the code cell with the `%use dataframe` line before any other code cells
> that use Kotlin DataFrame. This loads the library and makes APIs available

> > in the notebook. 

Reviewer 2: What do you think about:
Run the code cell with the `%use dataframe` line before any other code cells to make sure the library and its APIs are available in the notebook.
